Implements subpagination for ddgr

mosegontar opened the pull request:
Here's a provisional implementation of subpagination for ddgr. 

Pagination is set by using the `-n` (or `--num`) option which takes an integer N to paginate the results by. For example: `ddgr python -n 5` will paginate by 5 results at a time.

If the number of fetched results left to display is less than the pagination value, new results from the next Ddg results page are fetched and appended to the list of results already returned. Use of the `p` command in the omniprompt will show the previous N results, where N is the pagination value.

If the `f` command is used in the omni prompt, the index is reset to 0 so that the first set of results are displayed.

Please let me know if this is in line with what you are thinking/what changes should be made.

jarun commented:
I think we are on the same page. When we press `p`, we never fetch results, am I right?

jarun commented:
The indices do not start from 0. We have to hide (and store) the true first real index of the current page (as offset)) and start from 1 in each page. When users enters the _visible_ index to open a link in browser, we have to add that to current offset.

jarun commented:
Why don't we update the index when we add it to our _cache_?

jarun commented:
We try to fetch whenever we hit the last page. As we are caching the results, can we store the index (initialized to a negative value) to indicate which is our last index the first time we hit it? That way we can save the redundant fetch every time user reaches the last page.

jarun commented:
Self note: document the option in README, manpage, add to completion-scripts. (owner @jarun)

jarun commented:
It would be a good idea to default num to 10. 30/40 results per page needs scrolling up/down every time.

jarun commented:
Let n=0 disable num.

mosegontar commented:
I've made the following updates
- When paginating and fetching a new page of results, we update DdgCmd._urltable dict with the new set of results. The key for each result is equal to its index value + the previous total number of results retrieved. So if page 1 has 30 results, the 1st result on Page 2 will be added to the dict as 31.
-  Updated `do_open` and `cmdloop` to calculate correct index to open when paginating.
- Added a `display_index` parameter to Result.print() to set the displayed index value instead of the true index value.
- Set default --num value to 10.

There is a significant item remaining to be fixed: if pagination values are large, they can cause the index to increase to a value larger than the size of the results, and this causes problems with the output.

jarun commented:
Self-notes:

- I'll take up any review comments I might add.
- We need to fix the large `num` issue.
- Even if we are on the first page, `f`/`p` refreshes the list. Need to fix that.

jarun commented on ddgr:
```diff
@@ -1139,18 +1140,25 @@ class DdgCmd(object):
         parser.feed(page.text)
 
-        self.results = parser.results
+        if self.options.pagination:
```
Where is this getting set???

jarun commented on ddgr:
```diff
@@ -1267,8 +1290,15 @@ class DdgCmd(object):
     def do_next(self):
         # If no results were fetched last time, we have hit the last page already
-        if self._ddg_url._qrycnt == 0:
+        if self._ddg_url._qrycnt == 0 and self.index >= len(self.results):
```
What does this additional condition do???

jarun replied:
I think we will keep having these issues in deciphering stuff if we never get a single line of code comment.

jarun replied:
You add the check `self.options.pagination` in every other place and then fuse this kind of a condition without any check???

jarun replied:
Yes, I think this line is the issue with the _huge_ pagination value.

jarun replied:
I have _hidden_ it by limiting n to 25. If anyone is trying to set n greater than that he is trying to automate stuff. He should patch the code for himself. ;)

jarun commented on ddgr:
```diff
@@ -1139,18 +1140,25 @@ class DdgCmd(object):
+        if self.options.pagination:
+            self.results.extend(parser.results)
+            offset = len(self._urltable)
+            for r in self.results:
```
Are we doing this again and again for ALL results? Aren't the earlier results already indexed correctly?

jarun replied:
We are re-indexing results from offset??? Why???

jarun commented:
I'll probably have to do a massive rework on this. Anyway...

jarun commented on ddgr:
```diff
@@ -1161,19 +1169,25 @@ class DdgCmd(object):
                 sys.stderr.write(prelude)
-                for r in self.results:
-                    r.print()
+                for i, r in enumerate(results):
```
Broken for n=0

jarun commented:
I think we are reasonably good now! :+1:

jarun commented on ddgr:
```diff
@@ -1161,19 +1169,25 @@ class DdgCmd(object):
         else:
             # Regular output
-            if not self.results:
+            if not self.results or self.index > len(self.results):
```
Shouldn't the second check be `self.index > len(self.results`?

jarun replied:
Why do we need the second check? Please share an example with calculations.

jarun replied:
I think we can get rid of the check completely.
